### DISCUSSION TITLE
Partial redrawing

### DISCUSSION THREAD

**Discussion by semtexzv** [emilk/egui#913] [Q&A]
Hi, I'm just checking egui as a possible solution for my problem. My usecase is rather resource constrained. I need to create GUI for embedded application. In this environment, redrawing the whole screen is rather expensive (I'll be implementing a software renderer). Does egui currently support partial redraws ? eg:
* We have a 2d grid of boxes, each contains an animation, which only changes a few seconds.

In this case, it seems that it would be beneficial to re-render just the affected rectangles by sending following information to the backend:
* Clip rectangle
* Display list for the affected rectangle

**Comment by AlexApps99** [comment]
I don't think there's any functionality for damage tracking in egui, but you could probably implement that in your backend, by tracking changes to the meshes and what regions of the screen that would affect.
That does sound a bit convoluted, so if that doesn't work for your usecase, you might need to consider other options.
PRs are always welcome too

**Comment by emilk** [comment]
First of all: a software renderer for egui would be awesome! [I made a slow one for imgui](https://github.com/emilk/imgui_software_renderer) a while back, and have wanted to port it, but haven't gotten around to it. Please make sure to share yours!

There is nothing built-in for only rendering part of the screen, as all current egui backends assumes a fast hardware rasterizer.

I think it should be possible to add some caching in front of your software renderer. The [tessellator](https://docs.rs/egui/latest/egui/struct.Context.html#method.tessellate) batches draw calls by clipping rectangles. The batches roughly corresponds to panels, windows and scroll areas. So you could render each `ClippedMesh` to its own image buffer and composite the result. The next frame you hash the new `ClippedMesh` and check if you can reuse a buffer from the previous frame. Thus, if only one window or panel changes, then only that will be re-rendered.

If that is too coarse, then you can instead look at the `ClippedShape`:s returned by [`Context::end_frame`](https://docs.rs/egui/latest/egui/struct.Context.html#method.end_frame). You could break up the screen in cells (say 64x64 pixels big) and keep a set of hashes of all `ClippedShapes` that intersect that cell. At the start of each frame you see which cells each shape covers, and then you can see if a cell has any shapes added or removed. If so, that cell needs repainting.

**Comment by lambdachad** [comment]
Would it be possible to implement something like this to get best performance for eINK screens on Linux?
